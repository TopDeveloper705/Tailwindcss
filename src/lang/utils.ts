export function isSpace(str: string): boolean {
  return /\s/.test(str);
}

export function isDigit(str: string): boolean {
  return /^\d+$/.test(str);
}

export function isAlpha(str: string): boolean {
  return /\w/.test(str);
}

export function connect(strings: string[]): string {
  return strings.join(';\n') + ';';
}

export function findGroupEnd(text: string, startIndex = 0): number {
  let level = 1;
  let endBracket = searchFrom(text, '}', startIndex);
  while (endBracket !== -1) {
    const nextBracket = searchFrom(text, '{', startIndex);
    if (endBracket < nextBracket || nextBracket === -1) {
      level--;
      startIndex = endBracket + 1;
      if (level === 0) return endBracket;
    } else {
      level++;
      startIndex = nextBracket + 1;
    }
    endBracket = searchFrom(text, '}', startIndex);
  }
  return -1;
}

export function searchFrom(
  text: string,
  target: string | RegExp,
  startIndex = 0,
  endIndex?: number
): number {
  // search from partial of string
  const subText = text.substring(startIndex, endIndex);
  const relativeIndex = subText.search(target);
  return relativeIndex === -1 ? -1 : startIndex + relativeIndex;
}
