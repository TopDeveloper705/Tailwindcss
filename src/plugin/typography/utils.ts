export function uniq<T>(array: T[]): T[] {
  return Array.from(new Set(array));
}

export function castArray<T>(value: T): unknown[] {
  if (!arguments.length) {
    return [];
  }
  return Array.isArray(value) ? value : [value];
}

export function isUsableColor(color: string, values: string | {[key:string]:string}): boolean {
  return Boolean(values && typeof values === 'object' && color !== 'gray' && values[600]);
}

export const round = (num: number): string =>
  num
    .toFixed(7)
    .replace(/(\.[0-9]+?)0+$/, '$1')
    .replace(/\.0$/, '');

export const rem = (px: number): string => `${round(px / 16)}rem`;

export const em = (px: number, base: number): string => `${round(px / base)}em`;
