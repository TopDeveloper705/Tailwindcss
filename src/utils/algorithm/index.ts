export { default as sortSelector } from './sortSelector';
export { default as sortMediaQuery } from './sortMediaQuery';
export { default as compileStyleSheet } from './compileStyleSheet';
export { default as cssEscape } from './cssEscape';
export { default as combineConfig } from './combineConfig';
export { default as diffConfig } from './diffConfig';
