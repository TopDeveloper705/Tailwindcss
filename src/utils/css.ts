import { Property, Style, StyleSheet, InlineAtRule } from './style';
import { compile } from '../processor';
import screens from '../processor/variants/screen';

export class CSSParser {
    css:string;
    constructor(css:string) {
        this.css = css;
    }

    private _removeComment(css:string) {
        while (true) {
            const commentOpen = css.search(/\/\*/);
            const commentClose = css.search(/\*\//);
            if (commentOpen === -1 || commentClose === -1) break;
            css = css.substring(0, commentOpen) + css.substring(commentClose+2,);
        }
        return css;
    }

    private _searchGroup(text:string, startIndex=0) {
        let level = 1;
        while (true) {
            const endBracket = this._searchFrom(text, '}', startIndex);
            const nextBracket = this._searchFrom(text, '{', startIndex);
            if (endBracket === -1) return -1;
            if (endBracket < nextBracket || nextBracket === -1) {
                level --;
                startIndex = endBracket + 1;
                if (level == 0) return endBracket;
            } else {
                level ++;
                startIndex = nextBracket + 1;
            }
        }
    }

    private _searchFrom(text:string, target:string|RegExp, startIndex=0, endIndex=undefined) {
        // search from partial of string
        const subText = text.substring(startIndex, endIndex);
        const relativeIndex = subText.search(target);
        return relativeIndex === -1 ? -1 : startIndex + relativeIndex;
    }

    private _generateStyle(css:string, selector?:string, transform=false) {
        const properties:Property[] = [];
        const applies:string[] = [];
        let index = 0;
        while (true) {
            const start = this._searchFrom(css, /\S/, index);
            const end = this._searchFrom(css, ';', index);
            if (end === -1) break;
            const piece = css.substring(start, end+1);
            const isAtRule = css.charAt(start) === '@';
            const parsed = isAtRule? InlineAtRule.parse(piece) : Property.parse(piece);
            if (parsed) {
                if (transform && isAtRule && parsed.name==='apply' && parsed.value) {
                    applies.push(parsed.value);
                } else {
                    properties.push(parsed);
                }
            }
            index = end + 1;
        }
        if (applies.length > 0) {
            const styleSheet = compile(applies.join(' ')).styleSheet;
            styleSheet.children.forEach(style=>style.selector=selector);
            return (properties.length > 0) ? [new Style(selector, properties), ...styleSheet.children] : styleSheet.children;
        }
        return new Style(selector, properties);
    }

    private _clearDirectives(atrule:string):string|undefined {
        // transform @screen, keep @layer, remove other directives
        const iatrule = InlineAtRule.parse(atrule);
        if (["tailwind", "variants", "responsive"].includes(iatrule.name)) return undefined;
        if (iatrule.name === 'screen' && iatrule.value && screens.hasOwnProperty(iatrule.value)) return screens[iatrule.value]().atRules?.[0];
        return atrule;
    }

    parse(css=this.css, transform=false):StyleSheet {
        css = this._removeComment(css);
        let index = 0;
        const styleSheet = new StyleSheet();

        while (true) {
            const firstLetter = this._searchFrom(css, /\S/, index);
            if (firstLetter === -1) break;
            if (css.charAt(firstLetter) === '@') {
                // atrule
                const ruleEnd = this._searchFrom(css, ';', firstLetter);
                const nestStart = this._searchFrom(css, '{', firstLetter);

                if (nestStart === -1 || ruleEnd < nestStart) {
                    // inline atrule
                    const atrule = this._clearDirectives(css.substring(firstLetter, ruleEnd).trim());
                    if (atrule) styleSheet.add(InlineAtRule.parse(atrule).toStyle());
                    index = ruleEnd + 1;
                } else {
                    // nested atrule
                    const nestEnd = this._searchGroup(css, nestStart+1);
                    let atrule:string|undefined = css.substring(firstLetter, nestStart).trim();
                    if (transform) atrule = this._clearDirectives(atrule);
                    let result = this.parse(css.substring(nestStart + 1, nestEnd), transform).children.map(i=>i.atRule(atrule));
                    styleSheet.add(result);
                    index = nestEnd + 1;
                }
            } else {
                const nestStart = this._searchFrom(css, '{', firstLetter);
                const nestEnd = this._searchGroup(css, nestStart+1);
                if (nestStart === -1) {
                    // inline properties
                    styleSheet.add(this._generateStyle(css, undefined, transform));
                    break;
                }
                // nested selector
                styleSheet.add(this._generateStyle(css.substring(nestStart+1, nestEnd), css.substring(firstLetter, nestStart).trim(), transform));
                index = nestEnd + 1;
            }
        }
        return styleSheet;
    }
}