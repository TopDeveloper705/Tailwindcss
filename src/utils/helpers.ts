import { negateValue } from './tools';
import { Property, Style } from './style';
import type { FontSize, DarkModeConfig } from '../interfaces';

export function isString(value: unknown): value is string {
  return typeof value === 'string';
}

export function negative(scale: {
  [key: string]: string;
}): { [key: string]: string } {
  return Object.keys(scale)
    .filter((key) => scale[key] !== '0')
    .reduce(
      (negativeScale, key) => ({
        ...negativeScale,
        [`-${key}`]: negateValue(scale[key]),
      }),
      {}
    );
}

export function breakpoints(
  screens: { [key: string]: string } = {}
): { [key: string]: string } {
  return Object.keys(screens)
    .filter((key) => typeof screens[key] === 'string')
    .reduce(
      (breakpoints, key) => ({
        ...breakpoints,
        [`screen-${key}`]: screens[key],
      }),
      {}
    );
}

export function generateFontSize(font: FontSize): Property[] {
  if (typeof font === 'string') return [ new Property('font-size', font) ];
  const properties: Property[] = [];
  if (font[0]) properties.push(new Property('font-size', font[0]));
  if (typeof font[1] === 'string') {
    properties.push(new Property('line-height', font[1]));
  } else if (font[1]){
    if (font[1].lineHeight) properties.push(new Property('line-height', font[1].lineHeight));
    if (font[1].letterSpacing) properties.push(new Property('letter-spacing', font[1].letterSpacing));
  }
  return properties;
}

export function expandDirection(
  value: string,
  divide = false
): [a: string, b?: string] | undefined {
  const map: { [key: string]: [a: string, b?: string] } = {
    '': ['*'],
    y: ['top', 'bottom'],
    x: ['left', 'right'],
    t: divide ? ['top-left', 'top-right'] : ['top'],
    r: divide ? ['top-right', 'bottom-right'] : ['right'],
    b: divide ? ['bottom-right', 'bottom-left'] : ['bottom'],
    l: divide ? ['top-left', 'bottom-left'] : ['left'],
    tl: ['top-left'],
    tr: ['top-right'],
    br: ['bottom-right'],
    bl: ['bottom-left'],
  };
  if (value in map) return map[value];
}

export function generatePlaceholder(selector: string, property: Property | Property[], prefixer = false) : Style[] {
  if (!prefixer) return [ new Style(selector, property).pseudoElement('placeholder') ];
  return [
    new Style(selector, property).pseudoElement('-webkit-input-placeholder'),
    new Style(selector, property).pseudoElement('-moz-placeholder'),
    new Style(selector, property).pseudoClass('-ms-input-placeholder'),
    new Style(selector, property).pseudoElement('-ms-input-placeholder'),
    new Style(selector, property).pseudoElement('placeholder'),
  ];
}

export function toDarkStyle(style: Style, mode: DarkModeConfig): Style
export function toDarkStyle(style: Style[], mode: DarkModeConfig): Style[]
export function toDarkStyle(style: Style | Style[], mode: DarkModeConfig): Style | Style[] {
  if (!mode) return style;
  if (Array.isArray(style)) {
    if (mode === 'media') return style.map(i => new Style().atRule('@media (prefers-color-scheme: dark)').extend(i));
    return style.map(i => new Style().parent('.dark').extend(i));
  }
  if (mode === 'media') return new Style().atRule('@media (prefers-color-scheme: dark)').extend(style);
  return new Style().parent('.dark').extend(style);
}
