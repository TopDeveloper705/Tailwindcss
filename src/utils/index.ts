export * from "./tools";
export * from "./helpers";
