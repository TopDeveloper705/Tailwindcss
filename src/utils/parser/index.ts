export { default as HTMLParser } from './html';
export { default as CSSParser } from './css';
export { default as ClassParser } from './class';
export { default as SelectorParser } from './selector';
