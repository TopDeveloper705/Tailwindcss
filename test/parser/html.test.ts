import { readFileSync } from 'fs';
import { HTMLParser } from '../../src/utils/parser';

const EXAMPLE_1 = readFileSync('./test/assets/example.html').toString();
const EXAMPLE_2 = readFileSync('./test/assets/example2.html').toString();

describe('HTMLParser', () => {
  it('parse tags', () => {
    const parser = new HTMLParser();
    expect(parser.parseTags().length).toBe(0);
    parser.html = EXAMPLE_1;
    expect(parser.parseTags()).toMatchSnapshot('tag');
  });

  it('parse classes', () => {
    const parser = new HTMLParser();
    expect(parser.parseClasses().length).toBe(0);
    parser.html = EXAMPLE_1;
    expect(parser.parseClasses()).toMatchSnapshot('class');
  });

  it('parse attributes', () => {
    const parser = new HTMLParser(EXAMPLE_2);
    expect(parser.parseAttrs()).toMatchSnapshot('attr');
  });
});
