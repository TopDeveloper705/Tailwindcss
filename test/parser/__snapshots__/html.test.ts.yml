Tools / parse attributes / attr / 0: |-
  [
    {
      "raw": "container=\"~\"",
      "key": "container",
      "value": "~",
      "start": 5,
      "end": 18
    },
    {
      "raw": "min-h=\"screen\"",
      "key": "min-h",
      "value": "screen",
      "start": 19,
      "end": 33
    },
    {
      "raw": "bg=\"gray-100\"",
      "key": "bg",
      "value": "gray-100",
      "start": 34,
      "end": 47
    },
    {
      "raw": "flex=\"~ col\"",
      "key": "flex",
      "value": [
        "~",
        "col"
      ],
      "start": 48,
      "end": 60
    },
    {
      "raw": "justify=\"center\"",
      "key": "justify",
      "value": "center",
      "start": 61,
      "end": 77
    },
    {
      "raw": "p=\"y-6 sm:y-12 md:y-0 lg:y-2 xl:y-4\"",
      "key": "p",
      "value": [
        "y-6",
        "sm:y-12",
        "md:y-0",
        "lg:y-2",
        "xl:y-4"
      ],
      "start": 78,
      "end": 114
    },
    {
      "raw": "position=\"relative\"",
      "key": "position",
      "value": "relative",
      "start": 123,
      "end": 142
    },
    {
      "raw": "max-w=\"sm:xl\"",
      "key": "max-w",
      "value": "sm:xl",
      "start": 143,
      "end": 156
    },
    {
      "raw": "p=\"y-3\"",
      "key": "p",
      "value": "y-3",
      "start": 157,
      "end": 164
    },
    {
      "raw": "m=\"sm:x-auto\"",
      "key": "m",
      "value": "sm:x-auto",
      "start": 165,
      "end": 178
    },
    {
      "raw": "position=\"absolute inset-0\"",
      "key": "position",
      "value": [
        "absolute",
        "inset-0"
      ],
      "start": 189,
      "end": 216
    },
    {
      "raw": "gradient=\"to-r from-cyan-400 to-light-blue-500\"",
      "key": "gradient",
      "value": [
        "to-r",
        "from-cyan-400",
        "to-light-blue-500"
      ],
      "start": 217,
      "end": 264
    },
    {
      "raw": "shadow=\"lg\"",
      "key": "shadow",
      "value": "lg",
      "start": 265,
      "end": 276
    },
    {
      "raw": "transform=\"~ -skew-y-6 sm:skew-y-0 sm:-rotate-6\"",
      "key": "transform",
      "value": [
        "~",
        "-skew-y-6",
        "sm:skew-y-0",
        "sm:-rotate-6"
      ],
      "start": 277,
      "end": 325
    },
    {
      "raw": "border=\"sm:rounded-3xl\"",
      "key": "border",
      "value": "sm:rounded-3xl",
      "start": 326,
      "end": 349
    },
    {
      "raw": "position=\"relative\"",
      "key": "position",
      "value": "relative",
      "start": 371,
      "end": 390
    },
    {
      "raw": "p=\"x-4 y-10 sm:20\"",
      "key": "p",
      "value": [
        "x-4",
        "y-10",
        "sm:20"
      ],
      "start": 391,
      "end": 409
    },
    {
      "raw": "bg=\"white\"",
      "key": "bg",
      "value": "white",
      "start": 410,
      "end": 420
    },
    {
      "raw": "shadow=\"lg\"",
      "key": "shadow",
      "value": "lg",
      "start": 421,
      "end": 432
    },
    {
      "raw": "border=\"sm:rounded-3xl\"",
      "key": "border",
      "value": "sm:rounded-3xl",
      "start": 433,
      "end": 456
    },
    {
      "raw": "m=\"x-auto\"",
      "key": "m",
      "value": "x-auto",
      "start": 469,
      "end": 479
    },
    {
      "raw": "max-w=\"md\"",
      "key": "max-w",
      "value": "md",
      "start": 480,
      "end": 490
    },
    {
      "raw": "src=\"/img/logo.svg\"",
      "key": "src",
      "value": "/img/logo.svg",
      "start": 521,
      "end": 540
    },
    {
      "raw": "h=\"7 sm:8\"",
      "key": "h",
      "value": [
        "7",
        "sm:8"
      ],
      "start": 541,
      "end": 551
    },
    {
      "raw": "divide=\"y gray-200\"",
      "key": "divide",
      "value": [
        "y",
        "gray-200"
      ],
      "start": 582,
      "end": 601
    },
    {
      "raw": "text=\"base sm:lg gray-700\"",
      "key": "text",
      "value": [
        "base",
        "sm:lg",
        "gray-700"
      ],
      "start": 618,
      "end": 644
    },
    {
      "raw": "font=\"leading-6 sm:leading-7\"",
      "key": "font",
      "value": [
        "leading-6",
        "sm:leading-7"
      ],
      "start": 645,
      "end": 674
    },
    {
      "raw": "p=\"y-8\"",
      "key": "p",
      "value": "y-8",
      "start": 675,
      "end": 682
    },
    {
      "raw": "space=\"y-4\"",
      "key": "space",
      "value": "y-4",
      "start": 683,
      "end": 694
    },
    {
      "raw": "list=\"disc\"",
      "key": "list",
      "value": "disc",
      "start": 814,
      "end": 825
    },
    {
      "raw": "space=\"y-2\"",
      "key": "space",
      "value": "y-2",
      "start": 826,
      "end": 837
    },
    {
      "raw": "flex=~",
      "key": "flex",
      "value": "~",
      "start": 857,
      "end": 863
    },
    {
      "raw": "items=\"start\"",
      "key": "items",
      "value": "start",
      "start": 864,
      "end": 877
    },
    {
      "raw": "h=\"6 sm:7\"",
      "key": "h",
      "value": [
        "6",
        "sm:7"
      ],
      "start": 901,
      "end": 911
    },
    {
      "raw": "flex=\"~\"",
      "key": "flex",
      "value": "~",
      "start": 912,
      "end": 920
    },
    {
      "raw": "items=\"center\"",
      "key": "items",
      "value": "center",
      "start": 921,
      "end": 935
    },
    {
      "raw": "flex=\"shrink-0\"",
      "key": "flex",
      "value": "shrink-0",
      "start": 960,
      "end": 975
    },
    {
      "raw": "h=5",
      "key": "h",
      "value": "5",
      "start": 976,
      "end": 979
    },
    {
      "raw": "w=5",
      "key": "w",
      "value": "5",
      "start": 980,
      "end": 983
    },
    {
      "raw": "text=\"cyan-500\"",
      "key": "text",
      "value": "cyan-500",
      "start": 984,
      "end": 999
    },
    {
      "raw": "viewBox=\"0 0 20 20\"",
      "key": "viewBox",
      "value": [
        "0",
        "0",
        "20",
        "20"
      ],
      "start": 1000,
      "end": 1019
    },
    {
      "raw": "fill=\"currentColor\"",
      "key": "fill",
      "value": "currentColor",
      "start": 1020,
      "end": 1039
    },
    {
      "raw": "fill-rule=\"evenodd\"",
      "key": "fill-rule",
      "value": "evenodd",
      "start": 1067,
      "end": 1086
    },
    {
      "raw": "d=\"M10 18a8 8 0 100-16 8 8 0 000 16zm3.707-9.293a1 1 0 00-1.414-1.414L9 10.586 7.707 9.293a1 1 0 00-1.414 1.414l2 2a1 1 0 001.414 0l4-4z\"",
      "key": "d",
      "value": [
        "M10",
        "18a8",
        "8",
        "0",
        "100-16",
        "8",
        "8",
        "0",
        "000",
        "16zm3.707-9.293a1",
        "1",
        "0",
        "00-1.414-1.414L9",
        "10.586",
        "7.707",
        "9.293a1",
        "1",
        "0",
        "00-1.414",
        "1.414l2",
        "2a1",
        "1",
        "0",
        "001.414",
        "0l4-4z"
      ],
      "start": 1109,
      "end": 1246
    },
    {
      "raw": "clip-rule=\"evenodd\"",
      "key": "clip-rule",
      "value": "evenodd",
      "start": 1269,
      "end": 1288
    },
    {
      "raw": "class=\"ml-2\"",
      "key": "class",
      "value": "ml-2",
      "start": 1360,
      "end": 1372
    },
    {
      "raw": "class=\"text-sm font-bold text-gray-900\"",
      "key": "class",
      "value": [
        "text-sm",
        "font-bold",
        "text-gray-900"
      ],
      "start": 1433,
      "end": 1472
    },
    {
      "raw": "flex=\"~\"",
      "key": "flex",
      "value": "~",
      "start": 1563,
      "end": 1571
    },
    {
      "raw": "items=\"start\"",
      "key": "items",
      "value": "start",
      "start": 1572,
      "end": 1585
    },
    {
      "raw": "flex=\"~\"",
      "key": "flex",
      "value": "~",
      "start": 1609,
      "end": 1617
    },
    {
      "raw": "items=\"center\"",
      "key": "items",
      "value": "center",
      "start": 1618,
      "end": 1632
    },
    {
      "raw": "h=\"6 sm:7\"",
      "key": "h",
      "value": [
        "6",
        "sm:7"
      ],
      "start": 1633,
      "end": 1643
    },
    {
      "raw": "flex=\"shrink-0\"",
      "key": "flex",
      "value": "shrink-0",
      "start": 1668,
      "end": 1683
    },
    {
      "raw": "h=5",
      "key": "h",
      "value": "5",
      "start": 1684,
      "end": 1687
    },
    {
      "raw": "w=5",
      "key": "w",
      "value": "5",
      "start": 1688,
      "end": 1691
    },
    {
      "raw": "text=cyan-500",
      "key": "text",
      "value": "cyan-500",
      "start": 1692,
      "end": 1705
    },
    {
      "raw": "viewBox=\"0 0 20 20\"",
      "key": "viewBox",
      "value": [
        "0",
        "0",
        "20",
        "20"
      ],
      "start": 1706,
      "end": 1725
    },
    {
      "raw": "fill=\"currentColor\"",
      "key": "fill",
      "value": "currentColor",
      "start": 1726,
      "end": 1745
    },
    {
      "raw": "fill-rule=\"evenodd\"",
      "key": "fill-rule",
      "value": "evenodd",
      "start": 1773,
      "end": 1792
    },
    {
      "raw": "d=\"M10 18a8 8 0 100-16 8 8 0 000 16zm3.707-9.293a1 1 0 00-1.414-1.414L9 10.586 7.707 9.293a1 1 0 00-1.414 1.414l2 2a1 1 0 001.414 0l4-4z\"",
      "key": "d",
      "value": [
        "M10",
        "18a8",
        "8",
        "0",
        "100-16",
        "8",
        "8",
        "0",
        "000",
        "16zm3.707-9.293a1",
        "1",
        "0",
        "00-1.414-1.414L9",
        "10.586",
        "7.707",
        "9.293a1",
        "1",
        "0",
        "00-1.414",
        "1.414l2",
        "2a1",
        "1",
        "0",
        "001.414",
        "0l4-4z"
      ],
      "start": 1815,
      "end": 1952
    },
    {
      "raw": "clip-rule=\"evenodd\"",
      "key": "clip-rule",
      "value": "evenodd",
      "start": 1975,
      "end": 1994
    },
    {
      "raw": "m=l-2",
      "key": "m",
      "value": "l-2",
      "start": 2066,
      "end": 2071
    },
    {
      "raw": "font=\"bold\"",
      "key": "font",
      "value": "bold",
      "start": 2139,
      "end": 2150
    },
    {
      "raw": "text=\"sm gray-900\"",
      "key": "text",
      "value": [
        "sm",
        "gray-900"
      ],
      "start": 2151,
      "end": 2169
    },
    {
      "raw": "flex=\"~\"",
      "key": "flex",
      "value": "~",
      "start": 2243,
      "end": 2251
    },
    {
      "raw": "items=\"start\"",
      "key": "items",
      "value": "start",
      "start": 2252,
      "end": 2265
    },
    {
      "raw": "flex=\"~\"",
      "key": "flex",
      "value": "~",
      "start": 2289,
      "end": 2297
    },
    {
      "raw": "items=\"center\"",
      "key": "items",
      "value": "center",
      "start": 2298,
      "end": 2312
    },
    {
      "raw": "h=\"6 sm:7\"",
      "key": "h",
      "value": [
        "6",
        "sm:7"
      ],
      "start": 2313,
      "end": 2323
    },
    {
      "raw": "class=\"flex-shrink-0 h-5 w-5 text-cyan-500\"",
      "key": "class",
      "value": [
        "flex-shrink-0",
        "h-5",
        "w-5",
        "text-cyan-500"
      ],
      "start": 2348,
      "end": 2391
    },
    {
      "raw": "viewBox=\"0 0 20 20\"",
      "key": "viewBox",
      "value": [
        "0",
        "0",
        "20",
        "20"
      ],
      "start": 2392,
      "end": 2411
    },
    {
      "raw": "fill=\"currentColor\"",
      "key": "fill",
      "value": "currentColor",
      "start": 2412,
      "end": 2431
    },
    {
      "raw": "fill-rule=\"evenodd\"",
      "key": "fill-rule",
      "value": "evenodd",
      "start": 2459,
      "end": 2478
    },
    {
      "raw": "d=\"M10 18a8 8 0 100-16 8 8 0 000 16zm3.707-9.293a1 1 0 00-1.414-1.414L9 10.586 7.707 9.293a1 1 0 00-1.414 1.414l2 2a1 1 0 001.414 0l4-4z\"",
      "key": "d",
      "value": [
        "M10",
        "18a8",
        "8",
        "0",
        "100-16",
        "8",
        "8",
        "0",
        "000",
        "16zm3.707-9.293a1",
        "1",
        "0",
        "00-1.414-1.414L9",
        "10.586",
        "7.707",
        "9.293a1",
        "1",
        "0",
        "00-1.414",
        "1.414l2",
        "2a1",
        "1",
        "0",
        "001.414",
        "0l4-4z"
      ],
      "start": 2501,
      "end": 2638
    },
    {
      "raw": "clip-rule=\"evenodd\"",
      "key": "clip-rule",
      "value": "evenodd",
      "start": 2661,
      "end": 2680
    },
    {
      "raw": "m=\"l-2\"",
      "key": "m",
      "value": "l-2",
      "start": 2752,
      "end": 2759
    },
    {
      "raw": "text=\"base sm:lg\"",
      "key": "text",
      "value": [
        "base",
        "sm:lg"
      ],
      "start": 3009,
      "end": 3026
    },
    {
      "raw": "p=\"t-6\"",
      "key": "p",
      "value": "t-6",
      "start": 3027,
      "end": 3034
    },
    {
      "raw": "font=\"bold leading-6 sm:leading-7\"",
      "key": "font",
      "value": [
        "bold",
        "leading-6",
        "sm:leading-7"
      ],
      "start": 3035,
      "end": 3069
    },
    {
      "raw": "href=\"https://tailwindcss.com/docs\"",
      "key": "href",
      "value": "https://tailwindcss.com/docs",
      "start": 3157,
      "end": 3192
    },
    {
      "raw": "text=\"cyan-600 hover:cyan-700\"",
      "key": "text",
      "value": [
        "cyan-600",
        "hover:cyan-700"
      ],
      "start": 3193,
      "end": 3223
    }
  ]
Tools / parse classes / class / 0: |-
  [
    {
      "result": "container min-h-screen bg-gray-100 py-6 flex flex-col justify-center sm:py-12 md:py-0 lg:py-2 xl:py-4",
      "start": 12,
      "end": 113
    },
    {
      "result": "relative py-3 sm:max-w-xl sm:mx-auto",
      "start": 130,
      "end": 166
    },
    {
      "result": "absolute inset-0 bg-gradient-to-r from-cyan-400 to-light-blue-500 shadow-lg transform -skew-y-6 sm:skew-y-0 sm:-rotate-6 sm:rounded-3xl",
      "start": 185,
      "end": 320
    },
    {
      "result": "relative px-4 py-10 bg-white shadow-lg sm:rounded-3xl sm:p-20",
      "start": 350,
      "end": 411
    },
    {
      "result": "max-w-md mx-auto",
      "start": 432,
      "end": 448
    },
    {
      "result": "h-7 sm:h-8",
      "start": 507,
      "end": 517
    },
    {
      "result": "divide-y divide-gray-200",
      "start": 557,
      "end": 581
    },
    {
      "result": "py-8 text-base leading-6 space-y-4 text-gray-700 sm:text-lg sm:leading-7",
      "start": 606,
      "end": 678
    },
    {
      "result": "list-disc space-y-2",
      "start": 806,
      "end": 825
    },
    {
      "result": "flex items-start",
      "start": 853,
      "end": 869
    },
    {
      "result": "h-6 flex items-center sm:h-7",
      "start": 901,
      "end": 929
    },
    {
      "result": "flex-shrink-0 h-5 w-5 text-cyan-500",
      "start": 962,
      "end": 997
    },
    {
      "result": "ml-2",
      "start": 1366,
      "end": 1370
    },
    {
      "result": "text-sm font-bold text-gray-900",
      "start": 1439,
      "end": 1470
    },
    {
      "result": "flex items-start",
      "start": 1569,
      "end": 1585
    },
    {
      "result": "h-6 flex items-center sm:h-7",
      "start": 1617,
      "end": 1645
    },
    {
      "result": "flex-shrink-0 h-5 w-5 text-cyan-500",
      "start": 1678,
      "end": 1713
    },
    {
      "result": "ml-2",
      "start": 2082,
      "end": 2086
    },
    {
      "result": "text-sm font-bold text-gray-900",
      "start": 2162,
      "end": 2193
    },
    {
      "result": "flex items-start",
      "start": 2275,
      "end": 2291
    },
    {
      "result": "h-6 flex items-center sm:h-7",
      "start": 2323,
      "end": 2351
    },
    {
      "result": "flex-shrink-0 h-5 w-5 text-cyan-500",
      "start": 2384,
      "end": 2419
    },
    {
      "result": "ml-2",
      "start": 2787,
      "end": 2791
    },
    {
      "result": "pt-6 text-base leading-6 font-bold sm:text-lg sm:leading-7",
      "start": 3048,
      "end": 3106
    },
    {
      "result": "text-cyan-600 hover:text-cyan-700",
      "start": 3238,
      "end": 3271
    }
  ]
Tools / parse tags / tag / 0: |-
  [
    "div",
    "img",
    "p",
    "ul",
    "li",
    "span",
    "svg",
    "path",
    "code",
    "a"
  ]
